Desafio lambda3 - Nível intermediário - Olhar arquivo PDF nesta pasta com a descrição do desafio

In [2]:
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import classification_report

In [3]:
df = pd.read_csv('iris.csv')

In [4]:
df.head()

sepallength  sepalwidth  petallength  petalwidth        class
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepallength    150 non-null float64
sepalwidth     150 non-null float64
petallength    150 non-null float64
petalwidth     150 non-null float64
class          150 non-null object
dtypes: float64(4), object(1)
memory usage: 5.9+ KB


Receber as entradas específicas (variáveis pertinentes) e justificar a escolha das variáveis

In [18]:
X = df.drop('class',axis=1)

In [19]:
y = df['class']

In [22]:
#Transformando o atributo classe para valores numéricos utilizando o LabelEncoder
label = LabelEncoder()
label.fit(y)

LabelEncoder()

In [25]:
y = label.transform(y)

In [26]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [27]:
X.shape

(150, 4)

In [28]:
y.shape

(150,)

Algoritmo escolhido: K Nearest Neighbors (KNN) foi escolhido por sua simplicidade, rápida execução e por se tratar de uma base de dados pequena com poucos dados e atributos
<br>
Pensando na estratégia de resolver o problema proposto da melhor maneira e da forma mais simples possível, o algoritmo escolhido neste cenário apresentou bons resultados, sem a necessidade de tunning e ajustes

In [29]:
clf_knn = KNeighborsClassifier()

In [30]:
#Treinando o modelo utilizando a validação Cross Validation e armazenando os resultados
resultados_knn = cross_val_predict(clf_knn, X, y, cv=5)

In [31]:
#Verificando os resultados através de métricas
nomes_classes = label.inverse_transform([0,1,2])

In [34]:
nomes_classes

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

In [36]:
print(classification_report(y,resultados_knn,target_names=nomes_classes))

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        50
Iris-versicolor       0.98      0.94      0.96        50
 Iris-virginica       0.94      0.98      0.96        50

       accuracy                           0.97       150
      macro avg       0.97      0.97      0.97       150
   weighted avg       0.97      0.97      0.97       150



O modelo de classificação apresentou bons resultados utilizando a técnica de validação cruzada com um total de 5 folds
<br>
Foram escolhidas as métricas precision, recall e f1-score para avaliar o modelo por permitir uma análise por classe, já que estamos lidando com um problema que contém 3 classes.
<br>
Por meio da métrica de precisão podemos ver que o modelo sofreu mais para classificar a classe Iris-Virginica obtendo um percentual de 94% que corresponde ao número de vezes que o modelo classificou corretamente os dados da respectiva classe sobre a quantidade de vezes que a classe foi predita
<br>
Analise a métrica de revocação o modelo teve mais dificuldade para classificar os dados da classes iris-versicolor também com o percentual de 94% que corresponde ao número de vezes que o modelo classificou corretamente diante do volume de dados da respectiva classe no dataset
<br>
Por fim, a medida f1-score que corresponde a média harmônica entre as duas medidas anteriormente citadas apresentou um bom resultado com o valor de 97%

A métrica acurácia não se aplica neste caso, que é de classificação